pandas/_libs/tslibs/timezones.pyx

```diff
@@ -103,7 +103,24 @@ cpdef inline tzinfo maybe_get_tz(object tz):
             if isinstance(tz, _dateutil_tzfile) and '.tar.gz' in tz._filename:
                 tz._filename = zone
         else:
-            tz = pytz.timezone(tz)
+            if tz[0] == '-':
+                hours = int(tz[1:3])
+                minutes = int(tz[4:6])
+                tz = timezone(-timedelta(hours=hours, minutes=minutes))
+            elif tz[0] == '+':
+                hours = int(tz[1:3])
+                minutes = int(tz[4:6])
+                tz = timezone(timedelta(hours=hours, minutes=minutes))
+            elif tz[0:4] == 'UTC-':
+                hours = int(tz[4:6])
+                minutes = int(tz[7:9])
+                tz = timezone(-timedelta(hours=hours, minutes=minutes))
+            elif tz[0:4] == 'UTC+':
+                hours = int(tz[4:6])
+                minutes = int(tz[7:9])
+                tz = timezone(timedelta(hours=hours, minutes=minutes))
+            else:
```

Comment:
I meant you can remove the outer else, turn these all into elifs and dedent a level. There's also some code duplication you could limit with something like
```python
elif tz[0] in ["-", "+"]:
    delta = ...
    delta = - delta if tz[0] == "-" else delta
    return timezone(delta)
```

Reply:
Even better:
```python
elif tz[0] in {'-', '+'}:
    hours = int(tz[0:3])
    minutes = int(tz[4:6])
    tz = timezone(timedelta(hours=hours, minutes=minutes))
```

Reply:
This is not the same though. `timedelta(hours=-hours, minutes=minutes) != -timedelta(hours=hours, minutes=minutes)`. Again you'll want to add additional tests to cover all cases.

Reply:
When I add multiple idx values ```
idx = [
            datetime.datetime.now(datetime.timezone.utc),
            datetime.datetime.strptime(
                "2019-01-04T16:41:24+0200", "%Y-%m-%dT%H:%M:%S%z"
            ),
            datetime.datetime.strptime(
                "2019-01-04T16:41:24+0215", "%Y-%m-%dT%H:%M:%S%z"
            ),
            datetime.datetime.strptime(
                "2019-01-04T16:41:24-0200", "%Y-%m-%dT%H:%M:%S%z"
            ),
            datetime.datetime.strptime(
                "2019-01-04T16:41:24-0215", "%Y-%m-%dT%H:%M:%S%z"
            )
        ]``` I get a new error:
```
    def infer_object_encoding(data):
        head = data[:10] if isinstance(data, pd.Index) else data.dropna()[:10]
        if all(isinstance(i, STR_TYPE) for i in head) and not PY2:
            return "utf8"
        elif PY2 and all(isinstance(i, unicode) for i in head):
            return "utf8"
        elif PY2 and all(isinstance(i, (str, bytes)) for i in head):
            return "bytes"
        elif all(isinstance(i, bytes) for i in head):
            return 'bytes'
        elif all(isinstance(i, (list, dict)) for i in head):
            return 'json'
        elif all(isinstance(i, bool) for i in head):
            return 'bool'
        elif all(isinstance(i, integer_types) for i in head):
            return 'int'
        elif all(isinstance(i, float) or isinstance(i, np.floating)
                 for i in head):
            # You need np.floating here for pandas NaNs in object
            # columns with python floats.
            return 'float'
        else:
>           raise ValueError("Can't infer object conversion type: %s" % head)
E           ValueError: Can't infer object conversion type: 0    2019-01-04 16:41:24+02:00
E           1    2019-01-04 16:41:24+02:15
E           2    2019-01-04 16:41:24-02:00
E           3    2019-01-04 16:41:24-02:15
E           Name: index, dtype: object```

Reply:
Looks like the issue was that mixing timezones is not supported, so I converted it to a parametrized test.
Also fixed the `timedelta(hours=-hours, minutes=minutes) != -timedelta(hours=hours, minutes=minutes)` issue without branching, added tests for it. What do you think @dsaxton, is this a good direction?